include/faabric/scheduler/MpiThreadPool.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <map>
+#include <thread>
+#include <vector>
+
+#include <faabric/util/exception.h>
+#include <faabric/util/queue.h>
+
+namespace faabric::scheduler {
+// static faabric::util::Queue<std::pair<int, std::function<void (void)>>>
+// jobQueue;
+typedef faabric::util::Queue<std::pair<int, std::function<void(void)>>>
+  MpiJobQueue;
+
+class MpiAsyncThreadPool
+{
+  public:
+    explicit MpiAsyncThreadPool(int nThreads);
+
+    ~MpiAsyncThreadPool();
+
+    void awaitAsyncRequest(int reqId);
+
+    std::shared_ptr<MpiJobQueue> getMpiJobQueue();
+
+  protected:
+    void entrypoint(int i);
+
+  private:
+    std::vector<std::thread> threadPool;
+    std::mutex awakeMutex;
+    std::condition_variable awakeCV;
+
+    // TODO - use instead a data structure that is very easy to lookup. I am
```

Comment:
For the moment I've used a map (inherited from previous impl.) where:
* If `reqId` is a key it means that `reqId` has finished.
* If `reqId` is not a key, we must still wait.
Accesses to the map are protected by the `wait/notify` lock.

A map is not the right data structure for this as we don't use the value at all, and I may change this in the future. But this is not the current bottleneck.